---
title: "proteoM"
author:
- name: John Doe
date: "`r Sys.Date()`"
always_allow_html: true
output:
  html_document:
    fig_caption: yes
    highlight: haddock
    keep_md: yes
    theme: united
    toc: yes
    toc_depth: 4
    toc_float: yes
  md_document:
    toc: yes
    toc_depth: 4
    variant: gfm
  word_document:
    toc: yes
  github_document:
    toc: yes
  pdf_document:
    toc: yes
    number_sections: true
  html_vignette: 
  vignette: >
    %\VignetteIndexEntry{proteoM}
    %\VignetteEncoding{UTF-8}`
    %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  out.width='.40\\linewidth',
  fig.width=3,
  fig.height=3,
  fig.show='hold',
  fig.align='center',
  comment = "#>"
)
```


## Installation

To install this package, start R (latest version) and enter:

```{r installation, include = TRUE, eval = FALSE}
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("qzhang503/proteoM")
```

## Database searches

``` r
## An example of TMT-10plex
library(proteoQDA)

# Fasta databases 
copy_refseq_hs("~/proteoM/dbs/fasta/refseq")
copy_refseq_mm("~/proteoM/dbs/fasta/refseq")
copy_crap("~/proteoM/dbs/fasta/crap")

# MGF (e.g. by Proteome Discoverer)
# if with Thermo's `RawConverter (v1.2.0.1)`:
#   - Experiment Tpe: Data Dependent
#   - Options: Select monoisotopic m/z in DDA
copy_pd_mgf("~/proteoM/examples/mgfs")

# Ion searches
library(proteoM)

matchMS(
  out_path = "~/proteoM/examples", 
  mgf_path = "~/proteoM/examples/mgfs",
  fasta = c("~/proteoM/dbs/fasta/refseq/refseq_hs_2013_07.fasta", 
            "~/proteoM/dbs/fasta/refseq/refseq_mm_2013_07.fasta", 
            "~/proteoM/dbs/fasta/crap/crap.fasta"), 
  acc_type = c("refseq_acc", "refseq_acc", "other"), 
  fixedmods = c("TMT6plex (N-term)", "TMT6plex (K)", "Carbamidomethyl (C)"),
  varmods = c("Acetyl (Protein N-term)", "Oxidation (M)",
             "Deamidated (N)", "Gln->pyro-Glu (N-term = Q)"),
  max_miss = 4, 
  quant = "tmt10", 
  fdr_type = "protein", 
)


## An example of phospho TMT
library(proteoQDA)

# Try a different fasta database
copy_uniprot_hsmm("~/proteoM/dbs/fasta/uniprot")

# Try MGFs, e.g., by MSConvert
copy_msconv_mgf("~/proteoM/examples_p/mgfs")

# Ion searches
library(proteoM)

matchMS(
  out_path = "~/proteoM/examples_p", 
  mgf_path = "~/proteoM/examples_p/mgfs",
  fasta = c("~/proteoM/dbs/fasta/uniprot/uniprot_hsmm_2020_03.fasta", 
            "~/proteoM/dbs/fasta/crap/crap.fasta"), 
  acc_type = c("uniprot_acc", "other"), 
  fixedmods = c("TMT6plex (N-term)", "TMT6plex (K)", "Carbamidomethyl (C)"), 
  varmods = c("Acetyl (Protein N-term)", "Oxidation (M)", 
              "Deamidated (N)", "Phospho (S)", "Phospho (T)", 
              "Phospho (Y)", "Gln->pyro-Glu (N-term = Q)"), 
  max_miss = 4, 
  quant = "tmt10", 
  fdr_type = "protein", 
)
```

## Possible Next steps

- Search against real MGFs

  - With `RawConverter (v1.2.0.1)`:
    - Experiment Type
    - [x] Data Dependent
    - Options
    - [x] Select monoisotopic m/z in DDA
    - [x] Export Centroided Peaks in MS1
    - Output Formats
    - [x] MGF
  - With `MSConvert (v3.0.20287-769529fa4)`:
    - Options
    - [x] Output format: mgf
    - Filters
    - [x] peakPicking: msLevel=1-
- Data QC and informatics with [proteoQ](https://github.com/qzhang503/proteoQ/)


